src/fairseq2/models/mbart/builder.py

```diff
@@ -0,0 +1,295 @@
+
+from dataclasses import dataclass
+from typing import Literal, Optional
+
+from fairseq2.data import VocabularyInfo
+from fairseq2.models.transformer import (
+    TransformerEmbeddingFrontend,
+    TransformerFrontend,
+    TransformerModel,
+)
+from fairseq2.models.utils.arch_registry import ArchitectureRegistry
+from fairseq2.nn.embedding import Embedding
+from fairseq2.nn.position_encoder import (
+    LearnedPositionEncoder,
+    SinusoidalPositionEncoder,
+)
+from fairseq2.nn.projection import TiedProjection
+from fairseq2.nn.transformer import (
+    FeedForwardNetwork,
+    MultiheadAttention,
+    StandardFeedForwardNetwork,
+    StandardMultiheadAttention,
+    StandardTransformerDecoder,
+    StandardTransformerDecoderLayer,
+    StandardTransformerEncoder,
+    StandardTransformerEncoderLayer,
+    TransformerDecoder,
+    TransformerDecoderLayer,
+    TransformerEncoder,
+    TransformerEncoderLayer,
+    TransformerNormOrder,
+    create_default_sdpa,
+)
+from fairseq2.typing import DataType, Device
+
+
+@dataclass
+class mBartConfig:
+    """Holds the configuration of an mBart model."""
+
+    model_dim: int
+    """The dimensionality of the model."""
+
+    max_seq_len: int
+    """The expected maximum sequence length."""
+
+    vocabulary_size: int
+    """The size of the vocabulary."""
+
+    pad_idx: Optional[int]
+    """The index of the pad symbol in the vocabulary."""
+
+    num_encoder_layers: int
+    """The number of Transformer encoder layers."""
+
+    num_decoder_layers: int
+    """The number of Transformer decoder layers."""
+
+    num_encoder_attn_heads: int
+    """The number of attention heads in Transformer encoder layers."""
+
+    num_decoder_attn_heads: int
+    """The number of attention heads in Transformer decoder layers."""
+
+    ffn_inner_dim: int
+    """The inner dimensionality of Transformer feed-forward networks."""
+
+    # Position Encoder
+    pos_encoder_type: Literal["sinusoidal", "learned"]
+    """The type of position encoder."""
+
+    frontend_layernorm: bool
+    """Whether to add the layernorm in the encoder, decoder frontend."""
+
+    dropout_p: float
+    """The dropout probability in Transformer layers."""
+
+    norm_order: TransformerNormOrder
+    """The Layer Normalization order."""
+
+    def update_vocabulary(self, info: VocabularyInfo) -> None:
+        """Update vocabulary configuration from ``info``."""
+        self.vocabulary_size, self.pad_idx = info.size, info.pad_idx
+
+
+mbart_archs = ArchitectureRegistry[mBartConfig]("mbart")
+
+
+mbart_arch = mbart_archs.marker
+
+
+@mbart_arch("base")
+def _base() -> mBartConfig:
+    return mBartConfig(
+        model_dim=1024,
+        max_seq_len=1026,
+        vocabulary_size=65539,
+        pad_idx=0,
+        num_encoder_layers=12,
+        num_decoder_layers=12,
+        num_encoder_attn_heads=16,
+        num_decoder_attn_heads=16,
+        ffn_inner_dim=4096,
+        pos_encoder_type="learned",
```

Comment:
Looks like `pos_encoder_type` and `norm_order` are always `learned`, and `POST` according to [this](https://github.com/facebookresearch/fairseq/blob/main/fairseq/models/bart/model.py#L386). If that is the case, I would suggest removing these configuration parameters.

Reply:
I'm having to do this to successfully load the mBart checkpoint with UnitY: https://github.com/fairinternal/seamless_communication/pull/28/files#diff-189811785a49637a011c2db015430cfd708d92f832f8ef30ed7e10dc7f922635R103

The argument about `norm_order` makes sense, I'll remove that.